### ISSUE TITLE
FMRegressor predicts nans on Movielens dataset

### ISSUE THREAD

**Issue by crsanderford** [online-ml/river#669]
## Versions

**river version**: 0.7.1
**Python version**: 3.7.11
**Operating system**: Linux, Google Colab

## Describe the bug

Training a factorization machine model on River's included Movielens 100k dataset results in the trained FM predicting nan values. If the numerical side-features present in the dataset (timestamp, release date, age and zipcode) are excluded, the FMRegressor predicts properly, but including them breaks the regressor.

## Steps/code to reproduce

Here's a colab notebook that demonstrates the issue.

https://colab.research.google.com/drive/1MAIRIDTz1uBSTnkMe7jdIfF8QLoXFMKl?usp=sharing

**Comment by MaxHalford** [comment]
@gbolmier if you can take a look 🤗

**Comment by gbolmier** [comment]
Hey @crsanderford 👋  

What happened here is called *gradient explosion*, i.e. large error gradients accumulate and result in too large updates of the model weights, eventually leading to an overflow and the `nan` values you noticed.

Unfortunately, it often happens with this algorithm and numerical features. My guess is that it is because of the magnitude difference between feature values when you have categorical ones included between `[0,1]` and numerical ones in much larger intervals; these larger variations in numerical features having a huge impact on the predictions.

In order to make a model more stable it's common practice to transform numerical features into categorical ones, tuning the learning rate can also help. If you haven’t read it, I suggest you having a look at our *Examples* tab within the documentation, especially the [Matrix Factorization for Recommender Systems part 2](https://riverml.xyz/latest/examples/matrix-factorization-for-recommender-systems-part-2/) which have a [Feature engineering for FM models](https://riverml.xyz/latest/examples/matrix-factorization-for-recommender-systems-part-2/#feature-engineering-for-fm-models) section.

Cheers!

**Comment by MaxHalford** [comment]
Couldn't you also use a `StandardScaler`?

**Comment by gbolmier** [comment]
Not sure, but I think it's worth trying 👍

**Comment by crsanderford** [comment]
Hi @gbolmier,

Thanks for the help - just to be certain, you still think gradient explosion is the issue when the only numerical feature included is age? I suspected something to do with the size of the feature values when I saw how large the timestamps were, but the age values are obviously way smaller.

I had read the first of the recommender systems examples but no further - didn't realize there was one that pertained directly to FMs.

Thanks again!

**Comment by gbolmier** [comment]
You're welcome :) Actually could you try using a `StandardScaler` on your `age` feature as it was pointed out by Max?

**Comment by gbolmier** [comment]
The following works for me:

```python
from river import compose
from river import datasets
from river import evaluate
from river import facto
from river import metrics
from river import preprocessing

def evaluate_(model):
    X_y = datasets.MovieLens100K()
    metric = metrics.MAE() + metrics.RMSE()
    _ = evaluate.progressive_val_score(X_y, model, metric, print_every=25_000)

model = (
    compose.Select('user', 'item', 'occupation', 'gender') +
    (compose.Select('age') | preprocessing.StandardScaler()) |
    facto.FMRegressor(n_factors=10, seed=42)
)

evaluate_(model)
>>> [25,000] MAE: 0.776586, RMSE: 0.983164
>>> [50,000] MAE: 0.766496, RMSE: 0.970769
>>> [75,000] MAE: 0.765914, RMSE: 0.970504
>>> [100,000] MAE: 0.765622, RMSE: 0.969939
```

**Comment by crsanderford** [comment]
Can confirm that `StandardScaler` works for me as well.

Thanks again - closing the issue!

**Comment by MaxHalford** [comment]
In general you always want your features to be on the same scale. Even if the age don't look big, they're still on a different scale than binary features. Scaling is always a good idea.
